asf_heat_pump_suitability/pipeline/run_scripts/add_features.py

```diff
@@ -96,34 +96,30 @@ def main(epc_path: str, save_output: Optional[str] = None) -> pl.DataFrame:
     logging.info("Adding off gas grid column to EPC")
     off_gas_postcodes = off_gas.process_off_gas_data()
     enhanced_epc_df = off_gas.add_off_gas_feature(enhanced_epc_df, off_gas_postcodes)
-    logging.info("Adding Historic England listed buildings to EPC")
-    listed_buildings_england_df = listed_buildings.get_filtered_df_listed_buildings(
-        "E_historicengland_listed_buildings"
-    )
-    enhanced_epc_england_listed_buildings_df = (
-        listed_buildings.spatial_join_epc_with_listed_buildings(
-            enhanced_epc_df, listed_buildings_england_df
-        )
-    )
-    logging.info("Adding Cadw listed buildings to EPC")
-    listed_buildings_wales_df = listed_buildings.get_filtered_df_listed_buildings(
-        "W_cadw_listed_buildings"
-    )
-    enhanced_epc_wales_listed_buildings_df = (
-        listed_buildings.spatial_join_epc_with_listed_buildings(
-            enhanced_epc_df, listed_buildings_wales_df
-        )
+
+    # Add feature: listed buildings data
+    logging.info("Loading listed buildings for England")
+    e_listed_buildings_df = listed_buildings.transform_gdf_listed_buildings("England")
+    e_listed_buildings_df = listed_buildings.sjoin_df_epc_with_listed_buildings(
+        enhanced_epc_df, e_listed_buildings_df
     )
-    logging.info(
-        "Merging Historic England and Cadw listed buildings to a single column in EPC"
+
+    logging.info("Loading listed buildings for Wales")
+    w_listed_buildings_df = listed_buildings.transform_gdf_listed_buildings("Wales")
+    w_listed_buildings_df = listed_buildings.sjoin_df_epc_with_listed_buildings(
+        enhanced_epc_df, w_listed_buildings_df
     )
-    enhanced_epc_df = listed_buildings.merge_listed_buildings_nations(
-        enhanced_epc_england_listed_buildings_df, enhanced_epc_wales_listed_buildings_df
+
+    listed_buildings_df = pl.concat(
+        [e_listed_buildings_df, w_listed_buildings_df], how="vertical"
     )
+    enhanced_epc_df = enhanced_epc_df.join(listed_buildings_df, how="left", on="UPRN")
+
```

Comment:
Here we concat the inner joins of the listed buildings for England and Wales and then left join to EPC in one go. This avoids two bigger joins to the full EPC.